### ISSUE TITLE
Tree view only updates render when all nodes are closed and no nodes are selected 

### ISSUE THREAD

**Issue by jjpe** [EdJoPaTo/tui-rs-tree-widget#2]
I've started to experiment with the [provided example](https://github.com/EdJoPaTo/tui-rs-tree-widget/blob/main/examples/example.rs), and decided to expand it slightly to see if I could update data in the model, and have the view reflect that.

Below is the altered source code. 
`src/example.rs`:
```rust
fn main() -> Result<(), Box<dyn Error>> {
    // Terminal initialization
    let stdout = io::stdout().into_raw_mode()?;
    let stdout = MouseTerminal::from(stdout);
    let stdout = AlternateScreen::from(stdout);
    let backend = TermionBackend::new(stdout);
    let mut terminal = Terminal::new(backend)?;

    let events = Events::new();

    // App
    let mut app = App::new();

    loop {
        terminal.draw(|f| {
            let area = f.size();

            let items = Tree::new(app.tree.items.to_vec())
                .block(
                    Block::default()
                        .borders(Borders::ALL)
                        .title(format!("Tree Widget {:?}", app.tree.state)),
                )
                .highlight_style(
                    Style::default()
                        .fg(Color::Black)
                        .bg(Color::LightGreen)
                        .add_modifier(Modifier::BOLD),
                )
                .highlight_symbol(">> ");
            f.render_stateful_widget(items, area, &mut app.tree.state);
        })?;

        match events.next()? {
            Event::Tick => {},
            Event::Input(input) => match input {
                Key::Char('q') => { break },
                Key::Left  => { app.tree.close(); },
                Key::Right => { app.tree.open(); },
                Key::Down  => { app.tree.next(); },
                Key::Up    => { app.tree.previous(); },
                Key::Backspace => {
                    // println!("Foo");
                    let tree0 = StatefulTree::with_items(vec![
                        TreeItem::new_leaf("a"),
                        TreeItem::new(
                            "b",
                            vec![
                                TreeItem::new_leaf("c"),
                                TreeItem::new("d", vec![
                                    TreeItem::new_leaf("e"),
                                    TreeItem::new_leaf("f")
                                ]),
                                TreeItem::new_leaf("g"),
                            ],
                        ),
                        TreeItem::new_leaf("h"),
                    ]);
                    let tree1 = StatefulTree::with_items(vec![
                        TreeItem::new_leaf("aa"),
                        TreeItem::new(
                            "bb",
                            vec![
                                TreeItem::new_leaf("cc"),
                                TreeItem::new("dd", vec![
                                    TreeItem::new_leaf("ee"),
                                    TreeItem::new_leaf("ff")
                                ]),
                                TreeItem::new_leaf("gg"),
                            ],
                        ),
                        TreeItem::new_leaf("hh"),
                    ]);
                    let tree2 = StatefulTree::with_items(vec![
                        TreeItem::new_leaf("aaa"),
                        TreeItem::new(
                            "bbb",
                            vec![
                                TreeItem::new_leaf("ccc"),
                                TreeItem::new("ddd", vec![
                                    TreeItem::new_leaf("eee"),
                                    TreeItem::new_leaf("fff")
                                ]),
                                TreeItem::new_leaf("ggg"),
                            ],
                        ),
                        TreeItem::new_leaf("hhh"),
                    ]);

                    // NOTE: The comparisons below require a `PartialEq` impl for `StatefulTree`:
                    /***/if app.tree == tree0 { app.tree = tree1; }
                    else if app.tree == tree1 { app.tree = tree2; }
                    else if app.tree == tree2 { app.tree = tree0; }

                },
                _ => {},
            },
        }
    }

    Ok(())
}
``` 
`src/util/mod.rs`:
```rust
#[derive(Clone, PartialEq)] // NEW: derive the PartialEq trait. Also necessary for any type used by `StatefulTree`.
pub struct StatefulTree<'a> {
    pub state: TreeState,
    pub items: Vec<TreeItem<'a>>,
}
```

Specifically, what has changed is that the code now responds to pressing `Backspace` by updating the tree model data.
When `Backspace` is pressed, the code creates 3 `StatefulTree`s, and cycles through them depending on the model state at that time.

Armed with this example, I've noticed that the tree doesn't (fully) respond to the `Backspace` key event when any node has been selected, or when any node has been opened. I say fully because if I enable the `println!("Foo");` statement in the code, and open or select any node, the tree model update isn't rendered, but the `"Foo"` is indeed printed.

What I expected to happen instead was that the updated tree model would be reflected in the tree view the next time the tree is rendered.

**Comment by EdJoPaTo** [comment]
Thanks for your time and interest in this library!

I needed a moment to find the reason for this but its kinda interesting.
In the end you have the if else structure which checks `app.tree == tree0` which requires `PartialEq`. The first thing that bugged me was its incompleteness. There is an unhandled else case. So I added `else { panic!("bla") }`.
As you saw before when something is selected or opened nothing happened but now the panic was executed. That means something is interesting with that comparison of `PartialEq` as it sometimes does not match any of the if clauses.
Part of the `StatefulTree` is the `TreeState`. When you create a new `StatefulTree` the default state is assumed: Nothing is opened or selected. When you use the tui to select something the state changes. When comparing the default state to the new state with something selected they are different → the `app.tree == treeN` will be false in all 3 cases as the state will always be different.

I added a counter and checked its value instead of comparing `app.tree == treeN`:

```rust
counter += 1;
counter %= 3;

app.tree = if counter == 0 {
    tree0
} else if counter == 1 {
    tree1
} else {
    tree2
};
```

That worked fine in my testing.

I hope it helps and thanks again! :)
